### PR TITLE
Prometheus: Improve handling of special chars in label values

### LINKED ISSUE #96063: Prometheus: Handling of special characters in variables and label values
## Description

As described in https://github.com/grafana/grafana/pull/93415, there are some issues with the Prometheus data source's handling of variables that hold label values containing quotes. Queries either fail (in the case of double quotes) or return incorrect results (in the case of single quotes).

The same issue affects ad-hoc filters, where attempting to filter by label values containing special characters would generate invalid label matchers in the resulting PromQL query.

### How to reproduce

Start by spinning up the Prometheus instance here: https://github.com/NWRichmond/test-prometheus-quotes-and-special-chars, by following the instructions in that project's README.

Next, import the dashboard:

[Prometheus test special chars-1734045788113.json](https://github.com/user-attachments/files/18118649/Prometheus.test.special.chars-1734045788113.json)

### Actual results (current state)

Image

| Label value contains | Query expression sent to Prometheus                | Notes                                                                                                                                           |
|----------------------|----------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------|
| Single quote         | `"test_single_quote{label=\"looking\\\\'glass\"}"` | Doesn't result in a Prometheus error, but the expected results aren't returned.                                                                 |
| Double quote         | `"test_double_quote{label=\"\"quoted\"value\"}"`   | (Prometheus error) bad_data: invalid parameter "query": 1:34: parse error: unexpected identifier "glass" in label matching, expected "," or "}" |

### Expected results (aspirational state)

Image

| Label value contains | Query expression sent to Prometheus                |
|----------------------|----------------------------------------------------|
| Single quote         | `"test_single_quote{label=\"looking'glass\"}"`     |                              
| Double quote         | `"test_double_quote{label=\"looking\\\"glass\"}"`  |

### PR DESCRIPTION
**Pull request by NWRichmond**
## What is this change?

The Prometheus data source needed improvements in how it handles string escaping in different contexts.

- For regular label values, we need to properly escape double quotes and backslashes while leaving other characters unchanged. This is handled by the `prometheusRegularEscape` utility, which is updated by this PR.
- For regex patterns, we need more comprehensive escaping that handles double quotes, backslashes, and regex [metacharacters](https://simpleregex.dev/meta-characters) correctly. This is handled by the `prometheusSpecialRegexEscape` utility, which is updated by this PR.

This PR is similar to https://github.com/grafana/grafana/pull/93415, but focuses on ensuring that Explore Metrics works as expected. When #93415 was merged, we discovered that some values were being escaped multiple times or with too many backslashes, resulting in invalid queries being generated in Explore Metrics (example: `avg(some_metric{account=\\\"myaccount\\\"})`). This PR fixes that issue by preventing `prometheusRegularEscape` from adding escape characters to complete label matchers (such as `'account="myaccount"'`), as are used in Explore Metrics' filters.

To guard against unexpected regressions, all changes to escaping have been put behind a new `prometheusSpecialCharsInLabelValues` feature toggle.

This PR also aims to improve clarity in utilities & associated tests.

## Why do we need this change?

For the reasons described in #96063 and https://github.com/grafana/grafana/pull/93415.

## Who is this feature for?

Prometheus data source users.

## Which issue(s) does this PR fix?

Fixes #96063.

## Special notes for your reviewer

Be sure to enable the `prometheusSpecialCharsInLabelValues` feature toggle in `conf/custom.ini`. For example:
```
[feature_toggles]
enable = prometheusSpecialCharsInLabelValues
```

I'd appreciate your help confirming that the goals of #96063 and https://github.com/grafana/grafana/pull/93415 are achieved, and that the Explore Metrics experience is unaffected.

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: packages/grafana-prometheus/src/datasource.ts [hunk 1]
```diff
@@ -1041,13 +1045,49 @@ export function extractRuleMappingFromGroups(groups: RawRecordingRules[]): RuleQ
...
+  if (typeof value !== 'string') {
+    return value;
+  }
+
+  return (
+    value
+      // escape backslashes
+      .replace(/\\/g, '\\\\')
+      // escape single quotes with double backslashes
+      .replace(/'/g, "\\\\'")
```
Comment (gizmoguy) [inline packages/grafana-prometheus/src/datasource.ts:1061]: Single quotes don't need to be escaped here, but double quotes do:

```
// escape double quotes with single backslash
.replace(/"/g, '\\"')
```

The proposed code would escape non-regex promql queries like this:

```
singlequote_escape_test{label="looking\\'glass"}
doublequote_escape_test{label="looking"glass"}
```

The doublequote_escape_test query produces a syntax error from prometheus:

```
bad_data: invalid parameter "query": 1:40: parse error: unexpected identifier "glass" in label matching, expected "," or "}"
```

The correct queries should be:

```
singlequote_escape_test{label="looking'glass"} 1
doublequote_escape_test{label="looking\"glass"} 1
```

File: packages/grafana-prometheus/src/datasource.ts [hunk 2]
```diff
@@ -1041,13 +1045,49 @@ export function extractRuleMappingFromGroups(groups: RawRecordingRules[]): RuleQ
...
+  if (typeof value !== 'string') {
+    return value;
+  }
+
+  return (
+    value
+      // escape backslashes
+      .replace(/\\/g, '\\\\')
+      // escape double quotes
+      .replace(/"/g, '\\"')
```
Comment (gizmoguy) [inline packages/grafana-prometheus/src/datasource.ts:1091]: Single quotes don't need to be escaped here, but double quotes do:

```
// escape double quotes with triple backslashes
.replace(/"/g, '\\\\\\"')
```

The proposed code would escape regex promql queries like this:

```
singlequote_escape_test{label=~"looking\\'glass"}
doublequote_escape_test{label=~"looking"glass"}
```

The doublequote_escape_test query produces a syntax error from prometheus:

```
bad_data: invalid parameter "query": 1:36: parse error: unexpected identifier "glass" in label matching, expected "," or "}"
```

The correct queries should be:

```
singlequote_escape_test{label=~"looking'glass"} 1
doublequote_escape_test{label=~"looking\\\"glass"} 1
```

**Review by gizmoguy — CHANGES_REQUESTED** [review]
Hi @NWRichmond, I've spun up this PR against my dashboard I'm using for testing escaping, it throws a single quote, double quote and slashes at every prometheus query function.

My testing has revealed a couple of problems with how this PR handles double quotes which I've addressed in-line.

**Comment by NWRichmond** [comment]
Thank you very much for the review, @gizmoguy. I will make updates according to your feedback.

Before returning to work on this PR, I will be adding some end-to-end tests that validate some basic functionality related to [variables](https://grafana.com/docs/grafana/latest/dashboards/variables), the query builder, metrics explorer, etc. This will allow us to make changes to special character handling with confidence that we're not introducing any experience-breaking changes.

**Comment by gizmoguy** [comment]
Hi @NWRichmond 

I did some further testing on this PR last week, and realised if you incorporate my feedback on this PR, then we end up back in the same situation as we did in #93415, where dashboard queries with variables work, but explore metrics does not.

I had a bit of a look into how explore metrics builds queries using a query generator, and I see why this issue occurs now.

Variable interpolation for prometheus in grafana is used in a couple of different ways:

  1. Interpolate a variable (which could be regex or non-regex) which contains a single value into a a promql filter, e.g `$var1='looking"glass'` into metric{label="${var1}"}
  2. Interpolate a variable which containers a partial promql query (this is what explore metrics does) e.g `$filters='label="looking"glass"'` into `avg(${metric}{${filters}} ${otel_join_query})`

When the prometheus escaping functions (prometheusRegularEscape, prometheusSpecialRegexEscape) are called for case 1), they are escaping a value: e.g `looking"glass`. However for case 2) these escape functions are escaping a partial promql query: e.g `label="looking"glass"`. This is why the "over-escape" issue occurred for explore metrics when applying my PR #93415.

I've been thinking about how this might be fixable, however I'm not very familiar with the grafana internals. Ideally the query generator for explore metrics (or the interpolation functions) would be modified to treat escaping both case 1) and case 2) the same, i.e just the value would be escaped through the prometheus escaping functions (i.e `looking"glass`), rather than escaping the partial promql query in its entirety.

If this isn't possible, then I think a different set of escaping functions will be required to handle queries from a query generator for metrics explorer, as the escape requirements for both case 1) and 2) are different. I'm not sure it would be possible to implement these two different cases in the same escaping functions without a lot of regex, or a promql grammar parser that knows the difference between different parts of the promql query.

Interested to know your thoughts, and yes I agree that the E2E tests will help make changes to escaping less fragile in future.

**Comment by NWRichmond** [comment]
@gizmoguy thank you so much for digging into the Explore Metrics use case and providing such a cogent description of variable interpolation in the Prometheus data source. After returning to this PR with fresh eyes, I believe I've found a solution that satisfies the goals of https://github.com/grafana/grafana/issues/96063 and https://github.com/grafana/grafana/pull/93415. The PR description has been updated accordingly. Some time today, I will also provide some video demos and links to a reproduction. If you're able to give this another look, I'd really appreciate your feedback.

**Review by gizmoguy — APPROVED** [review]
Hi @NWRichmond 

Glad my comments were useful, and nice work on this patch.

I have tested with the `prometheusSpecialCharsInLabelValues` feature enabled against my internal test dashboard and played around with explore metrics and haven't been able to find any functional issues handling special characters.

Code also looks good, so this gets a LGTM from me :+1: 

escape_test_dashboard

**Comment by NWRichmond** [comment]
@gizmoguy amazing! Thank you so much for testing and sharing your findings.

**Comment by NWRichmond** [comment]
@itsmylife here are some things that should make it easier to review this PR:

- Run [this Prometheus instance](https://github.com/NWRichmond/test-prometheus-quotes-and-special-chars) and use it in a new data source (note that it runs on `localhost:9094` and requires no auth)
- Add the dashboard provided in #96063: [Prometheus test special chars-1734045788113.json](https://github.com/user-attachments/files/18118649/Prometheus.test.special.chars-1734045788113.json)
- Play around with the dashboard. Every label value provided by variables should work, and any ad-hoc label value should work.
- Play around with the new data source in Explore Metrics. Panels should always work, regardless of which label values are added to the filters.

Here are my findings:

## Demo

### Dashboard panels

https://github.com/user-attachments/assets/7e9a3348-73fc-4e47-a33d-531d841187e1

### Explore Metrics

https://github.com/user-attachments/assets/cce222ba-c948-444b-82c1-77d626f603fd

**Comment by itsmylife** [comment]
Dashbaords and all variables are working good. Great job 💪 
But, OTel is not working properly in Explore metrics. And I think this is a problem in Explore metrics. I think we need to check how we use OTel variables in Explore Metrics. 

image

**Comment by itsmylife** [comment]
@NWRichmond I did a change which will fix the issue in Explore Metrics with OTel. With the changes I made there is no more problems. Please you also check my changes. 
@bohandley Could you please take a look at this too?

**Comment by NWRichmond** [comment]
@itsmylife I tested Explore Metrics with the OTel experience both enabled & disabled, and everything is working as expected ✨ 

I noticed that there was one place we were missing some guarding via the feature toggle - that's been fixed now (https://github.com/grafana/grafana/pull/96067/commits/664817664b5e0d57f5b3508d8fa9d3167dd063b4).

**Review by itsmylife — APPROVED** [review]
🎉

**Comment by NWRichmond** [comment]
Just noticed that special chars support isn't working yet in Builder Mode. Converting to Draft until we get that sorted out.

**Review by bohandley — APPROVED** [review]
LGTM! Tested with OTel and with the feature toggle we are not getting the previous errors.

**Comment by NWRichmond** [comment]
@itsmylife @bohandley FYI I snuck in two more changes:
- https://github.com/grafana/grafana/pull/96067/commits/a9d865774d618b41f9135628c98d569cf94a1980: In code mode, autocomplete now works when there are special chars in label values: 

https://github.com/user-attachments/assets/444529b4-caad-4553-9fe4-450e96fd3ad8
- in https://github.com/grafana/grafana/pull/96067/commits/f5dd96930f6894c0281209d92fa1419e6f822c50, I removed the use of `prometheusRegularEscape` when constructing `labelFilters` in `packages/grafana-prometheus/src/datasource.ts`. After the changes in https://github.com/grafana/grafana/pull/96067/commits/1c3d0ae3f183a70212362ada8b1455ad83a5b954, we no longer need this.

I tested both of these changes with:
- Explore Metrics
  - with OTel Experience
  - without OTel Experience
- Classic Explore
  - code mode
  - builder mode
- Dashboards
  - using variables, as shown in the demo above
 
Everything continues to work the way I'd expect.

**Review by itsmylife — APPROVED** [review]

File: public/app/features/trails/MetricSelect/api.ts [hunk 3]
```diff
@@ -1,12 +1,15 @@
 import { AdHocVariableFilter, RawTimeRange, Scope } from '@grafana/data';
 import { getPrometheusTime } from '@grafana/prometheus/src/language_utils';
+import { PromQueryModeller } from '@grafana/prometheus/src/querybuilder/PromQueryModeller';
```
Comment (NWRichmond) [inline public/app/features/trails/MetricSelect/api.ts:3]: yet another thing that we'll need to export from `@grafana/prometheus` when Explore Metrics is externalized

File: public/app/features/trails/MetricSelect/api.ts [hunk 4]
```diff
@@ -99,3 +104,8 @@ export async function getMetricNamesWithScopes(
     missingOtelTargets: false,
   };
 }
+
+function removeBrackets(input: string): string {
+  const match = input.match(/^\{(.*)\}$/); // extract the content inside the brackets
+  return match?.[1] ?? '';
+}
```
Comment (itsmylife) [inline public/app/features/trails/MetricSelect/api.ts:111]: In my utf8 support PR I did same kind of thing you did here. Let's see how they'll play together.

File: public/app/features/trails/MetricSelect/api.ts [hunk 5]
```diff
@@ -31,7 +34,9 @@ export async function getMetricNamesWithoutScopes(
   instances: string[],
   limit?: number
 ) {
-  const matchTerms = adhocFilters.map((filter) => `${filter.key}${filter.operator}"${filter.value}"`);
+  const matchTerms = adhocFilters.map((filter) =>
+    removeBrackets(queryModeller.renderLabels([{ label: filter.key, op: filter.operator, value: filter.value }]))
+  );
```
Comment (itsmylife) [inline public/app/features/trails/MetricSelect/api.ts:39]: Let's put this behind the feature flag too

Comment (NWRichmond) [reply public/app/features/trails/MetricSelect/api.ts]: Good idea, done in https://github.com/grafana/grafana/pull/96067/commits/7cf6c3cd4cfaa09dc081bfeaace03dfd287957f4 ✅

**Comment by NWRichmond** [comment]
I've tested this extensively with `prometheusSpecialCharsInLabelValues` enabled and disabled, in the following scenarios:

- Explore Metrics
  - applying single & multiple label values to filters
    - with & without OTel Experience
    - in Metric Select and Metric scenes
- Classic Explore
  - code mode
    - with autocompleted label values
  - builder mode
  - with **Explain** enabled
- Dashboards
  - using variables, as shown in the demo above
  - using ad-hoc filters for label values, as shown in the demo above

After seeing the expected results, I feel confident about merging this PR.

**Pull request merged**